### ISSUE TITLE
opt: invalid datum type given: int, expected interval

### ISSUE THREAD

**Issue by justinj** [cockroachdb/cockroach#35968]
Was trying to repro a different panic but found this one.

cc @mjibson in case you've seen this one before and this is a dup. Suspiciously similar to #35113.

```
  CREATE TABLE table0 (
      col0 TIMESTAMP NULL,
      col1 OID NULL,
      col2 UUID NULL,
      col3 OID[] NULL,
      col4 TIMESTAMP NULL,
      col5 INTERVAL NULL,
      UNIQUE INDEX table0_col1_col0_col4_key (col1 ASC, col0 ASC, col4 ASC),
      INDEX table0_col1_col2_idx (col1 ASC, col2 ASC),
      INDEX table0_col0_idx (col0 ASC),
      INDEX table0_col0_col4_idx (col0 DESC, col4 DESC),
      UNIQUE INDEX table0_col5_col2_col0_key (col5 ASC, col2 ASC, col0 ASC),
      UNIQUE INDEX table0_col5_col4_col0_col1_col2_key (col5 ASC, col4 DESC, col0 DESC, col1 DESC, col2 ASC),
      INDEX table0_col5_idx (col5 ASC),
      FAMILY fam0 (col4, rowid),
      FAMILY fam1 (col0, col3, col2, col5, col1)
  );

DELETE FROM
    sqlsmith.public.table0 AS tab_147
WHERE
    NULL
    IN (
            SELECT
                NULL AS col_303
            FROM
                [
                    UPDATE
                        sqlsmith.public.table0 AS tab_148
                    SET
                        col3
                            = tab_148.col1::OID
                            || CASE
                                WHEN NULL THEN ARRAY[1819542824, 2203341712, 2649628927]
                                ELSE ARRAY[
                                    1574174933,
                                    2263982923,
                                    3853361441,
                                    696729079,
                                    855335641,
                                    2496945150
                                ]
                                END::OID[]
                    LIMIT
                        23:::INT8
                    RETURNING
                        col0::TIMESTAMP - col5::INTERVAL AS ret_79,
                        (
                            WITH
                                with_36
                                    (
                                        col_291,
                                        col_292,
                                        col_293,
                                        col_294,
                                        col_295,
                                        col_296,
                                        col_297,
                                        col_298,
                                        col_299,
                                        col_300,
                                        col_301
                                    )
                                    AS (
                                        SELECT
                                            *
                                        FROM
                                            (
                                                VALUES
                                                    (
                                                        NULL,
                                                        NULL,
                                                        false,
                                                        current_date()::DATE,
                                                        '228.81.139.165/26':::INET,
                                                        e'\\\\xb1c9085a'::BYTES
                                                        || e'\\\\x774b530b1eee313cd0'::BYTES,
                                                        '772530c8-43a8-4e50-9a8d-25ecd47b9a0d':::UUID,
                                                        '"z_iG404e>@"':::JSONB,
                                                        'e65f:df50:a69c:a01b:4ef9:d14:eaca:f1fd/75':::INET,
                                                        (-0.12293548799596238):::FLOAT8,
                                                        NULL
                                                    ),
                                                    (
                                                        true,
                                                        log(
                                                            7.220375248374971069E+33::DECIMAL
                                                        )::DECIMAL,
                                                        true,
                                                        current_date()::DATE,
                                                        '95de:a455:d4ab:f836:d228:cb39:2975:e1fb/24':::INET,
                                                        e'\\\\x6112915f682b5952':::BYTES,
                                                        crdb_internal.cluster_id()::UUID,
                                                        jsonb_object(
                                                            ARRAY[
                                                                'I|l>K9',
                                                                '6029j51d$0f'
                                                            ]::STRING[],
                                                            ARRAY['']::STRING[]
                                                        )::JSONB,
                                                        '65.33.255.173/20':::INET,
                                                        mod(
                                                            1.4550765109075012::FLOAT8,
                                                            (
                                                                1.6405813311841413::FLOAT8
                                                                / (-0.557762836547297):::FLOAT8::FLOAT8
                                                            )::FLOAT8
                                                        )::FLOAT8::FLOAT8
                                                        / 0.07760265588785842::FLOAT8,
                                                        '63f3b470-18a0-4a06-a284-61939ebe3d45':::UUID
                                                    ),
                                                    (
                                                        true,
                                                        NULL,
                                                        false,
                                                        NULL,
                                                        hostmask(
                                                            set_masklen(
                                                                '1ca8:67a2:4ba8:e685:8ba:538f:b188:cf79/100'::INET,
                                                                (-2565734613004299742):::INT8::INT8
                                                            )::INET::INET
                                                        )::INET,
                                                        e'\\\\x2be5c1e28c':::BYTES,
                                                        '0b330424-eb25-4830-8e13-98338416b9c4':::UUID,
                                                        '{"#vv{Vf@M<sS": false, ".8G": null, ">#J+LN4g0*": [], "OSs3/ecKSxT": [0.3982822680377671, []], "baz": {}}':::JSONB,
                                                        NULL,
                                                        (-1.0859384285401368):::FLOAT8,
                                                        'd2925962-3333-4066-9dc4-35165ccedc5f':::UUID
                                                    ),
                                                    (
                                                        false,
                                                        (-6630142.562240724481):::DECIMAL,
                                                        false,
                                                        '1971-02-20':::DATE,
                                                        '53.203.156.21/9':::INET,
                                                        e'\\\\x':::BYTES,
                                                        'a4644aa5-6657-43ce-882d-1d3430a0c367':::UUID,
                                                        '{"_s9": [[[{}]]], "b": {"!VK+": false}, "bar": [{"baz": true}, true], "c": true, "uKrU,": [{"foo": []}, {}]}'::JSONB,
                                                        'e53a:7a95:c2db:b86c:967c:ea57:17bd:c800/13':::INET,
                                                        (-2.0869471302750804):::FLOAT8,
                                                        NULL
                                                    )
                                            )
                                                AS tab_149
                                                    (
                                                        col_291,
                                                        col_292,
                                                        col_293,
                                                        col_294,
                                                        col_295,
                                                        col_296,
                                                        col_297,
                                                        col_298,
                                                        col_299,
                                                        col_300,
                                                        col_301
                                                    )
                                    )
                            SELECT
                                tab_150.col0 AS col_302
                            FROM
                                sqlsmith.public.table0 AS tab_150,
                                sqlsmith.public.table0 AS tab_151,
                                sqlsmith.public.table0 AS tab_152,
                                with_36,
                                sqlsmith.public.table0 AS tab_153
                            LIMIT
                                1:::INT8
                        )
                            AS ret_80,
                        true AS ret_81
                ],
                sqlsmith.public.table0 AS tab_154,
                sqlsmith.public.table0 AS tab_155,
                sqlsmith.public.table0 AS tab_156
            WHERE
                ret_81
        )
ORDER BY
    tab_147.col0 ASC, tab_147.col1 ASC
LIMIT
    1:::INT8
```

```
goroutine 280 [running]:
runtime/debug.Stack(0x7454d20, 0xc000b15940, 0xc000000003)
        /usr/local/Cellar/go/1.11/libexec/src/runtime/debug/stack.go:24 +0xa7
github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x7454d20, 0xc000b15940, 0xc00063f300, 0x6b5d420, 0xc006011140, 0x1)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:226 +0xa6
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00552a000, 0x7454d20, 0xc000b15940, 0x67f1b20, 0xc005f338b0)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:707 +0x2dd
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00552a000, 0x7454d20, 0xc000b15940)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:426 +0x61
panic(0x67f1b20, 0xc005f338b0)
        /usr/local/Cellar/go/1.11/libexec/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:172 +0x25d
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc005fdaf00, 0x0, 0x0, 0x0, 0x6)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:186 +0x54e
github.com/cockroachdb/cockroach/pkg/sql.(*rowSourceToPlanNode).Next(0xc005fe6480, 0x7454de0, 0xc005fc2ae0, 0xc005fd4c00, 0xc00552a350, 0x7, 0x0, 0x0)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/row_source_to_plan_node.go:78 +0x63
github.com/cockroachdb/cockroach/pkg/sql.(*spoolNode).startExec(0xc005fc2750, 0x7454de0, 0xc005fc2ae0, 0xc005fd4c00, 0xc00552a350, 0xc005fe6480, 0xc005380bb8)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/spool.go:62 +0x18b
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2(0x6ce3244, 0x5, 0x7456e20, 0xc005fc2750, 0xc000faa720, 0xc005380bc8)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:495 +0x55
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1(0xc005f558c0, 0x6ce3244, 0x5, 0x7456e20, 0xc005fc2750)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:145 +0x5d
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc005f558c0, 0x7456e20, 0xc005fc2750, 0x6ce3244, 0x5)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:615 +0x1c4
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc005f558c0, 0x7456e20, 0xc005fc2750, 0x203001, 0x40043a3)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(0x7454de0, 0xc005fc2ae0, 0x7456e20, 0xc005fc2750, 0x0, 0x6e51ea0, 0x0, 0x0, 0x0, 0xc005fc2ba0, ...)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:76 +0x1a4
github.com/cockroachdb/cockroach/pkg/sql.startExec(0x7454de0, 0xc005fc2ae0, 0xc005fd4c00, 0xc00552a350, 0x7456e20, 0xc005fc2750, 0xc005381a30, 0x2)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:498 +0x116
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc005fdb400, 0x7454de0, 0xc005fc2ae0, 0x0, 0xc0005ff680)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:124 +0xd9
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*hashJoiner).Start(0xc005fe8000, 0x7454de0, 0xc005fc2ae0, 0x92e8580, 0x67f0da0)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/hashjoiner.go:222 +0x52
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc005fe8000, 0x7454de0, 0xc005fc2ae0)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:800 +0x52
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc005fd23c0, 0x7454de0, 0xc005fc2ae0, 0x6e515a8, 0x0, 0x0)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:627 +0x1e9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000458500, 0xc005f554a0, 0xc005ed38c0, 0xc005382768, 0xc0057962c0, 0xc005fd4800, 0x0)
        /Users/justin/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:252 +0x8ad
```

**Comment by mjibson** [comment]
Cool I haven't seen that. I think I need to add all datum types to my simple test table.

**Comment by justinj** [comment]
Fixed by #36375
